truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedCallTarget.java

```diff
@@ -881,6 +939,10 @@ public final boolean compile(boolean lastTierCompilation) {
                     return true;
                 }
 
+                if (isCompilationStopped()) {
```

Comment:
I would move this check right before the synchronized block. Doing this in the synchronized block does not give us any advantage as its global anyway, but it is costly as its done on every call target call in the interpreter.